File: leetcode004/main.go
package main

import (
	"fmt"
)

func main() {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	fmt.Println(maximumSubarray(nums))
}

func maximumSubarray(nums []int) int {
	// Define maxSub to store sum of values in a sub array
	maxSub, currentSum := nums[0], 0
	// Loop through all values in the nums array
	for _, num := range nums {
		// If currentSum is negative, ignore it, it will reduce the current sum
		if currentSum < 0 {
			currentSum = 0
		}
		currentSum += num
		// Get the max sum
		maxSub = max(maxSub, currentSum)
	}

	return maxSub
}

func max(num1, num2 int) int {
	// Returns the max value between two values
	if num1 > num2 {
		return num1
	} else {
		return num2
	}
}

File: leetcode004/main.py
from typing import List

def maxSubArray(nums:List[int]) -> int :
    # Get the total sum
    sum = nums[0]
    # Define current sum
    curr_sum = 0

    # Iterate through values and get the sum
    for num in nums :
        # if curr_sum sum is negative value
        if curr_sum < 0 :
            curr_sum = 0
        # Calculates the current sum
        curr_sum += num
        # Get the maximum value between curr_sum and sum
        sum = max(curr_sum, sum)
    return sum


if __name__ == "__main__":
    numbers = [-2,1,-3,4,-1,2,1,-5,4]
    results = maxSubArray(numbers)
    print(results)